### PR TITLE
CAPI callback error signaling

### PR DESCRIPTION
**Pull request by atobiszei**
* Fix bug in async execution
Additionally there are few improvements in building:
* stop recompiling whole OVMS with dependencies when enabling OV_TRACING
* further ovms_lib separation which should enhance further using bazel cache and rebuilds
Eg. changing from python/mp disable to enable should affect lesser scope of compilation units which should improve bazel caching.

Ticket:CVS-141785

### 🧪 Checklist

- [x] Unit tests added.
- [x] The documentation updated.
- [x] Change follows security best practices.

### CODE REVIEW

File: src/capi_frontend/capi.cpp [hunk 1]
```diff
@@ -929,6 +929,43 @@ static Status getPipelineDefinition(Server& server, const std::string& servableN
...
+        CallbackGuard(OVMS_InferenceRequestCompletionCallback_t userCallback, void* userCallbackData, OVMS_InferenceResponse** userResponse, std::unique_ptr<ovms::InferenceResponse>&& response) :
+            userCallback(userCallback),
+            userCallbackData(userCallbackData),
+            userResponsePtr(userResponse),
+            response(std::move(response)) {}
+        ~CallbackGuard() {
+            if (!success) {
+                if (!userCallback)
+                   return;
+                SPDLOG_DEBUG("Calling user provided callback with success: {}", success);
```
Comment (dkalinowski) [inline src/capi_frontend/capi.cpp:947]: this log will not be printed with any other value than success=false anyway

File: src/modelinstance.cpp [hunk 2]
```diff
@@ -1611,47 +1611,81 @@ Status ModelInstance::inferAsync(const RequestType* requestProto,
...
+                    }
+                }
+                std::unique_ptr<ResponseType> res(new ResponseType(this->getName(), this->getVersion()));
+                OutputGetter<ov::InferRequest&> outputGetter(inferRequest);
                 try {
-                    SPDLOG_DEBUG("rethrow_exception");
-                    std::rethrow_exception(exception);
-                } catch (const std::exception& e) {
-                    SPDLOG_DEBUG("got exception in ov::InferRequest callback: {}", e.what());
+                    // TODO created filter based on what is in request, then perform casual serialization for what was NOT in request, and rewrite tensors from request to response for those that were
```
Comment (dkalinowski) [inline src/modelinstance.cpp:1675]: what do you mean exactly? do you mean output filtering via output name selection inside request?

File: src/test/c_api_tests.cpp [hunk 3]
```diff
@@ -1921,3 +1844,98 @@ TEST(CAPI, MultipleThreadsStarting) {
...
+        return ovms::StatusCode::OK;
+    }
+    void setOutputsInfo(const tensor_map_t& outputsInfo) {
+        this->outputsInfo = outputsInfo;
+    }
+};
+
+const float INITIAL_VALUE{0.13666};
+const float GARBAGE_VALUE = 42.66613;
+const float FLOAT_TOLLERANCE{0.001};
```
Comment (dkalinowski) [inline src/test/c_api_tests.cpp:1885]:
```suggestion
const float FLOAT_TOLERANCE{0.001};
```

File: src/modelinstance.cpp (hunk 2 above)
Comment (atobiszei) [reply src/modelinstance.cpp:1675]: This is old TODO:
https://github.com/openvinotoolkit/model_server/pull/2801/files/38b759c2a21106cbeab7fa9e77181324047b0b9e#diff-ea20c94d98c26b6ddf560dae96db24b9d293c8a7d75418e744666c52dfa8553cL1642
This idea for slight optimization

File: src/test/c_api_tests.cpp (hunk 3 above)
Comment (atobiszei) [reply src/test/c_api_tests.cpp]: Fixed as well in other tests

**Review by dkalinowski — APPROVED** [review]

File: src/modelinstance.cpp [hunk 4]
```diff
@@ -1611,47 +1611,81 @@ Status ModelInstance::inferAsync(const RequestType* requestProto,
...
+                        userCallbackData(userCallbackData),
+                        request(request) {}
+                    ~CallbackGuard() {
+                        SPDLOG_DEBUG("Calling user provided callback with success: {}", success);
+                        if (!success) {
+                            userCallback(nullptr, 1, userCallbackData);
+                        } else {
+                            userCallback(response, 0, userCallbackData);
+                        }
+                        SPDLOG_DEBUG("Called user provided callback");
```
Comment (michalkulakowski) [inline src/modelinstance.cpp:1643]: Is this log necessary? If so, maybe log level could be changed to trace

**Review by michalkulakowski — APPROVED** [review]

**Pull request merged**